### PR TITLE
feat(cloud): add stripe checkout, customer portal, and usage metering

### PR DESCRIPTION
**Pull request by marcklingen**

### CODE REVIEW

File: web/src/ee/features/billing/server/stripeWebhookApiHandler.ts [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+import { env } from "@/src/env.mjs";
+import { type NextRequest, NextResponse } from "next/server";
+import { prisma } from "@langfuse/shared/src/db";
+import * as Sentry from "@sentry/node";
+import { stripeClient } from "@/src/ee/features/billing/utils/stripe";
+import type Stripe from "stripe";
+import { CloudConfigSchema, parseDbOrg } from "@langfuse/shared";
+
+const STRIPE_WEBHOOK_SIGNING_SECRET_DEV: string | null =
+  "whsec_12dc385262f1a5d0f4ba1507cc81f9b3a3e2d03fd99d4ad625b8e21c87dcfd37";
```
Comment (maxdeichmann) [inline web/src/ee/features/billing/server/stripeWebhookApiHandler.ts:14]: looks like dev but should this be in here?

File: web/src/ee/features/billing/server/stripeWebhookApiHandler.ts [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+  const orgId = getOrgIdFromStripeClientReference(clientReference);
+  const subscriptionId = session.subscription; // Assuming subscription ID is in session.subscription
+
+  // check if org exists and that the customer and subscription IDs match
+  const organization = await prisma.organization.findUnique({
+    where: { id: orgId },
+  });
+  if (!organization) {
+    console.error("[Stripe Webhook] Organization not found");
+    Sentry.captureMessage("[Stripe Webhook] Organization not found", {
```
Comment (maxdeichmann) [inline web/src/ee/features/billing/server/stripeWebhookApiHandler.ts:137]: Please use the otel integration now

`traceException` from `import { traceException } from "@langfuse/shared/src/server";`

File: web/src/ee/features/billing/server/stripeWebhookApiHandler.ts [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+    },
+  });
+}
+
+async function handleSubscriptionChanged(
+  subscription: Stripe.Subscription,
+  action: "created" | "deleted" | "updated",
+) {
+  // wait for 3 seconds to ensure the subscription is fully created during the checkout session
+  await new Promise((resolve) => setTimeout(resolve, 3000));
```
Comment (maxdeichmann) [inline web/src/ee/features/billing/server/stripeWebhookApiHandler.ts:207]: lol

File: web/src/ee/features/billing/server/cloudBillingRouter.ts [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+import { parseDbOrg } from "@langfuse/shared";
+import {
+  createTRPCRouter,
+  protectedOrganizationProcedure,
+} from "@/src/server/api/trpc";
+import { TRPCError } from "@trpc/server";
+import * as z from "zod";
+import { throwIfNoOrganizationAccess } from "@/src/features/rbac/utils/checkOrganizationAccess";
+
+export const cloudBillingRouter = createTRPCRouter({
```
Comment (maxdeichmann) [inline web/src/ee/features/billing/server/cloudBillingRouter.ts:15]: I dont see any audit logs here. I think we should add these.

File: worker/src/queues/cloudUsageMeteringQueue.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+export const cloudUsageMeteringJobExecutor =
+  redis && env.STRIPE_SECRET_KEY
+    ? new Worker(
+        QueueName.CloudUsageMeteringQueue,
+        async (job) => {
+          if (job.name === QueueJobs.CloudUsageMeteringJob) {
+            return instrumentAsync(
+              { name: "cloudUsageMeteringJobExecutor" },
```
Comment (maxdeichmann) [inline worker/src/queues/cloudUsageMeteringQueue.ts:40]: Shall i add this to dd dashboard or will you do this?

File: web/src/ee/features/billing/server/stripeWebhookApiHandler.ts (hunk 1 above)
Comment (marcklingen) [reply web/src/ee/features/billing/server/stripeWebhookApiHandler.ts:14]: yes, will be removed before going live, is non critical (dev mode only)

File: web/src/ee/features/billing/server/stripeWebhookApiHandler.ts (hunk 3 above)
Comment (marcklingen) [reply web/src/ee/features/billing/server/stripeWebhookApiHandler.ts]: https://github.com/langfuse/langfuse/pull/2962/commits/4dc6b5be58582ba5eca722662934b0a0b865bf71

File: web/src/ee/features/billing/server/cloudBillingRouter.ts (hunk 4 above)
Comment (marcklingen) [reply web/src/ee/features/billing/server/cloudBillingRouter.ts:15]: done for create checkout session

File: worker/src/queues/cloudUsageMeteringQueue.ts (hunk 5 above)
Comment (marcklingen) [reply worker/src/queues/cloudUsageMeteringQueue.ts:40]: it'd be great if you can do dashboard then

File: web/src/ee/features/billing/server/stripeWebhookApiHandler.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcklingen) [reply web/src/ee/features/billing/server/stripeWebhookApiHandler.ts]: fixed

**Pull request merged**